### PR TITLE
first AAVQE added

### PR DESCRIPTION
**Pull request by nicolezattarin**
A basic model for Adiabatically Assisted VQE

### CODE REVIEW

**Comment by scarrazza** [comment]
@nicolezattarin thank you for the implementation.
I would like to debug the github action, could you please try to commit some change, e.g. use a single empty line between class methods?

File: src/qibo/models/variational.py [hunk 1]
```diff
@@ -97,6 +97,142 @@ def _loss(params, circuit, hamiltonian):
...
+        return s
+
+    def hamiltonian(self, t):
+        """Returns the adiabatic evolution Hamiltonian at a given time."""
+        if (t - self._t_max) > self.ATOL:
+            raise_error(ValueError, "t cannot be greater than {}, but it is {}.".format(self._t_max, t))
+        # boundary conditions  s(0)=0, s(total_time)=1
+        st = self.schedule(t)
+        return self._h0 * (1 - st) + self._h1 * st
+
```
Comment (scarrazza) [inline src/qibo/models/variational.py:203]: Please removing this empty space.

File: src/qibo/models/variational.py [hunk 2]
```diff
@@ -97,6 +97,142 @@ def _loss(params, circuit, hamiltonian):
...
+    Example:
+    ::
+        import numpy as np
+        from qibo import gates, models, hamiltonians
+        # create circuit ansatz for two qubits
+        circuit = models.Circuit(2)
+        circuit.add(gates.RY(0, theta=0))
+        circuit.add(gates.RY(1, theta=0))
+        # define the easy and the problem Hamiltonian.
```
Comment (scarrazza) [inline src/qibo/models/variational.py:121]:
```suggestion
        # define the easy and the problem Hamiltonians.
```

File: src/qibo/models/variational.py [hunk 3]
```diff
@@ -97,6 +97,142 @@ def _loss(params, circuit, hamiltonian):
...
+        easy_hamiltonian=hamiltonians.X(2)
+        problem_hamiltonian=hamiltonians.XXZ(2)
+        # define a scheduling function with only one parameter 
+        # and boundary conditions s(0) = 0, s(1) = 1
+        s = lambda t: t
+        # create AAVQE model 
+        aavqe = models.AAVQE(circuit, easy_hamiltonian, problem_hamiltonian, 
+                             s, nsteps = 10, Tmax = 1)
```
Comment (scarrazza) [inline src/qibo/models/variational.py:129]:
```suggestion
                             s, nsteps=10, Tmax=1)
```

File: src/qibo/models/variational.py [hunk 4]
```diff
@@ -97,6 +97,142 @@ def _loss(params, circuit, hamiltonian):
...
+                                   "object but is {}.".format(type(easy_hamiltonian)))
+        if type(problem_hamiltonian) != type(easy_hamiltonian):
+            raise_error(TypeError, "The problem Hamiltonian should be of the same type of the easy Hamiltonian"
+                                    " ({}), but it is {}.".format(type(easy_hamiltonian), type(problem_hamiltonian)))
+        if easy_hamiltonian.nqubits != problem_hamiltonian.nqubits:
+            raise_error(ValueError, "The easy Hamiltonian has {} qubits while problem Hamiltonian has {}."
+                                    "".format(easy_hamiltonian.nqubits, problem_hamiltonian.nqubits))
+
+        self.ATOL = 1e-7 # Tolerance for checking s(0) = 0 and s(T) = 1.
+        self.ATOL_TIME = 1e-7 # Tolerance for checking if time is greater than t_max.
```
Comment (scarrazza) [inline src/qibo/models/variational.py:150]: Please move both attributes to new arguments (with default value of `1e-7`) in the `__init__` function.

**Comment by scarrazza** [comment]
Thanks @nicolezattarin, could you please edit the `.github/workflows/rules.yml` file and add:
```diff
diff --git a/.github/workflows/rules.yml b/.github/workflows/rules.yml
index d64c103d..5c5befa3 100644
--- a/.github/workflows/rules.yml
+++ b/.github/workflows/rules.yml
@@ -1,7 +1,7 @@
 # A single CI script with github workflow
 name: Tests
 
-on: [push]
+on: [push, pull_request_target]
 
 jobs:
   build:
```
and commit and push?

File: src/qibo/models/variational.py [hunk 5]
```diff
@@ -97,6 +97,142 @@ def _loss(params, circuit, hamiltonian):
         return result, parameters, extra
 
 
+class AAVQE(object):
+    """This class implements the Adiabatically Assisted Variational Quantum Eigensolvers
```
Comment (bpcarlos) [inline src/qibo/models/variational.py:101]:
```suggestion
    """This class implements the Adiabatically Assisted Variational Quantum Eigensolver
```

File: src/qibo/models/variational.py [hunk 6]
```diff
@@ -97,6 +97,142 @@ def _loss(params, circuit, hamiltonian):
...
+    algorithm. See https://arxiv.org/abs/1806.02287.
+
+    Args:
+        circuit (:class:`qibo.abstractions.circuit.AbstractCircuit`): variaional ansatz.
+        easy_hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): initial Hamiltonian object.
+        problem_hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): problem Hamiltonian object.
+        s (callable): scheduling function of time that defines the scheduling of the
+            adiabatic evolution. It must verify boundary contitions: s(0) = 0 and s(1) = 1.
```
Comment (bpcarlos) [inline src/qibo/models/variational.py:109]:
```suggestion
            s (callable): scheduling function of time that defines the adiabatic
            evolution. It must verify boundary contitions: s(0) = 0 and s(1) = 1.
```

File: src/qibo/models/variational.py [hunk 7]
```diff
@@ -97,6 +97,142 @@ def _loss(params, circuit, hamiltonian):
...
+        nsteps (float): number of steps of the adiabatic evolution.
+        t_max (float): total time of time evolution.
```
Comment (bpcarlos) [inline src/qibo/models/variational.py:111]:
```suggestion
        t_max (float): total time evolution.
```

File: .github/workflows/rules.yml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 # A single CI script with github workflow
 name: Tests
 
-on: [push]
+on: [push, pull_request_target]
```
Comment (scarrazza) [inline .github/workflows/rules.yml:4]: Thanks, but I think it doesn't work. Could you please replace the `pull_request_target` with `pull_request`?

**Comment by scarrazza** [comment]
@nicolezattarin perfect, it works, could you please add the same on `.github/workflows/wheels.yml`?

**Comment by scarrazza** [comment]
@nicolezattarin, it looks like there are some pylint issues here, could you please have a look at this report: https://github.com/qiboteam/qibo/pull/475/checks?check_run_id=3782265076

**Comment by AdrianPerezSalinas** [comment]
Hi all, 
I have reviewed the code, it seems to work well. I have made tests and they are fine. I implemented some changes.
1) in the `vaiational.AAVQE` model, I deleted a `schedule` function that was repeated
2) I have written a new `main.py` file to accommodate the tutorial to the new scheme

There is something I could not solve. These lines 
image
are not working for me, and it looks like it has to do with the version of qibo. 
image (1)

However, it is the last one I have installed. Is there any key problem here?

**Comment by nicolezattarin** [comment]
Hi, I will review that piece of code today, also because there are pylint issues related to that. I have to check if there are changes in the model for adiabatic evolution

**Comment by AdrianPerezSalinas** [comment]
By the way, have you performed benchmarks on the AAVQE?

**Comment by scarrazza** [comment]
In order to fix the coverage, I think should be sufficient to include a simple test for this new class in `src/qibo/tests/test_models_variational.py`. @nicolezattarin if you have some time could you please try to include a new test by using the `test_vqe` as template?

**Comment by scarrazza** [comment]
@nicolezattarin thanks. I believe there is just an argument issue [here](https://github.com/qiboteam/qibo/pull/475/checks?check_run_id=3801083198#step:9:356).

**Comment by nicolezattarin** [comment]
I have the results: tests are performed optimising the Hamiltonian of the Heisenberg XXZ model. The ansatz is the same of the example [here](https://qibo.readthedocs.io/en/stable/code-examples/tutorials/aavqe/README.html). The main outcome is that the AAVQE approach is helpful as the number of qubits is increased, in particular on gradient-free methods such as COBYLA and Nelder-Mead in this case. For what concern a classic BFGS the approach does not provide advantages in terms of convergence, while requires a huge execution time if compared with a classical VQE.


2qubits_aavqe_benchmarks
4qubits_aavqe_benchmarks
6qubits_aavqe_benchmarks

**Comment by AdrianPerezSalinas** [comment]
Hi all, 

these results are something expectable, I guess that the advantage comes as many qubits are considered. However, I would like to drive your attention to something. Is there any control over the number of measurements? I would expect to have better results with AAVQE with the same number of measurements. Do we have some control on that?

**Comment by nicolezattarin** [comment]
Hi all,
In this simulation the number of iterations is not fixed, I can test it if you think it is proper to do that. Anyway, I would expect the advantage of AAVQE to be more evident as well, since at the end we would perform optimizer_iter*AAVQE_iter iterations.

**Comment by scarrazza** [comment]
@nicolezattarin could you please merge the latest master in this branch?

**Comment by nicolezattarin** [comment]
Merged

**Comment by mlazzarin** [comment]
I will try to reproduce the issue with python 3.6.

**Comment by nicolezattarin** [comment]
Updated, thank you

File: src/qibo/models/variational.py [hunk 9]
```diff
@@ -96,6 +96,135 @@ def _loss(params, circuit, hamiltonian):
         return result, parameters, extra
 
 
+class AAVQE(object):
+    """This class implements the Adiabatically Assisted Variational Quantum Eigensolver
+    algorithm. See https://arxiv.org/abs/1806.02287.
+
+    Args:
+        circuit (:class:`qibo.abstractions.circuit.AbstractCircuit`): variaional ansatz.
```
Comment (mlazzarin) [inline src/qibo/models/variational.py:104]:
```suggestion
        circuit (:class:`qibo.abstractions.circuit.AbstractCircuit`): variational ansatz.
```

File: src/qibo/models/variational.py [hunk 10]
```diff
@@ -96,6 +96,135 @@ def _loss(params, circuit, hamiltonian):
...
+        easy_hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): initial Hamiltonian object.
+        problem_hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): problem Hamiltonian object.
+        s (callable): scheduling function of time that defines the adiabatic
+            evolution. It must verify boundary contitions: s(0) = 0 and s(1) = 1.
```
Comment (mlazzarin) [inline src/qibo/models/variational.py:108]:
```suggestion
            evolution. It must verify boundary conditions: s(0) = 0 and s(1) = 1.
```

File: src/qibo/models/variational.py [hunk 11]
```diff
@@ -96,6 +96,135 @@ def _loss(params, circuit, hamiltonian):
...
+        nsteps (float): number of steps of the adiabatic evolution.
+        t_max (float): total time evolution.
+        bounds_tolerance (float): tolerance for checking s(0) = 0 and s(T) = 1.
```
Comment (mlazzarin) [inline src/qibo/models/variational.py:111]: Is it ``s(T)=1`` or ``s(1)=1``?

File: src/qibo/models/variational.py [hunk 12]
```diff
@@ -96,6 +96,135 @@ def _loss(params, circuit, hamiltonian):
...
+        time_tolerance (float): tolerance for checking if time is greater than t_max.
+
+    Example:
```
Comment (mlazzarin) [inline src/qibo/models/variational.py:114]: We should add this to the doctests, see #487 .

File: src/qibo/models/variational.py [hunk 13]
```diff
@@ -96,6 +96,135 @@ def _loss(params, circuit, hamiltonian):
...
+
+        if nsteps <= 0: # pragma: no cover
+            raise_error(ValueError, "Number of steps nsteps should be positive but is {}."
+                                    "".format(nsteps))
+        if t_max <= 0: # pragma: no cover
+            raise_error(ValueError, "Maximum time t_max should be positive but is {}."
+                                    "".format(t_max))
+        if easy_hamiltonian.nqubits != problem_hamiltonian.nqubits: # pragma: no cover
+            raise_error(ValueError, "The easy Hamiltonian has {} qubits while problem Hamiltonian has {}."
+                                    "".format(easy_hamiltonian.nqubits, problem_hamiltonian.nqubits))
```
Comment (mlazzarin) [inline src/qibo/models/variational.py:147]: We may also implement some tests for the errors and remove ``#  pragma: no cover`` from these if statements, but maybe it's excessive. The same holds for all the other checks.

File: src/qibo/models/variational.py [hunk 14]
```diff
@@ -96,6 +96,135 @@ def _loss(params, circuit, hamiltonian):
...
+            jac (dict): Method for computing the gradient vector for scipy optimizers.
+            hess (dict): Method for computing the hessian matrix for scipy optimizers.
+            hessp (callable): Hessian of objective function times an arbitrary
+                            vector for scipy optimizers.
+            bounds (sequence or Bounds): Bounds on variables for scipy optimizers.
+            constraints (dict): Constraints definition for scipy optimizers.
+            tol (float): Tolerance of termination for scipy optimizers.
+            options (dict): a dictionary with options for the different optimizers.
+            compile (bool): whether the TensorFlow graph should be compiled.
+            processes (int): number of processes when using the paralle BFGS method.
```
Comment (mlazzarin) [inline src/qibo/models/variational.py:214]:
```suggestion
            processes (int): number of processes when using the parallel BFGS method.
```

**Review by mlazzarin — COMMENTED** [review]
Hi! It looks like the issue with python36 is caused by the new aavqe tests, but I don't know why and I cannot reproduce it.

Btw I left some very minor comments regarding this PR, let me know what do you think.
@scarrazza do I have push access to this branch?

**Comment by scarrazza** [comment]
@mlazzarin thanks, if you are using qibojit please give a try with `NUMBA_NUM_THREADS=1 python test_script.py`.

**Comment by mlazzarin** [comment]
> @mlazzarin thanks, if you are using qibojit please give a try with `NUMBA_NUM_THREADS=1 python test_script.py`.

I tried but I still cannot reproduce the issue. I guess it appears only with a restricted number of configurations. Moreover, the CI uses python 3.6.15, but I can't find a way to install it by using ``pip`` or ``conda``, so I'm using 3.6.13. So I'm probably using a different environment.

**Comment by AdrianPerezSalinas** [comment]
Since we have already decided to not maintain python 3.6, we can merge this branch. 

I have however a last-minute question. Do we have a documentation like in the VQE page? I loaded the sphinx documentation but could not find it

**Comment by scarrazza** [comment]
@AdrianPerezSalinas, I will disable py36 for the CI in another PR.
Concerning the sphinx docs, the VQE is in https://qibo.readthedocs.io/en/stable/api-reference/qibo.html#variational-quantum-eigensolver-vqe

**Comment by AdrianPerezSalinas** [comment]
Okay, but there is no exclusive AAVQE page. I will add it and then we can merge

**Comment by AdrianPerezSalinas** [comment]
I' have seen that it is done but not compiled for some reason, so we can merge and doc will be updated  consequently

**Pull request merged**